File: ios/SignatureView.swift
// Ink algorithm ported from signature_pad (https://github.com/szimek/signature_pad),
// MIT License, Copyright (c) 2023 Szymon Nowak.
// Keep the math in sync with src/signatureMath.ts — it is validated by the
// golden vectors in test-fixtures/signature-golden.json.

import ExpoModulesCore
import UIKit

struct InkSample {
  var x: CGFloat
  var y: CGFloat
  /** Milliseconds; only deltas matter. */
  var time: Double
}

class SignatureView: ExpoView {
  // MARK: - Ink parameters (set via props)

  var penColor: UIColor = .black
  var minWidth: CGFloat = 0.5
  var maxWidth: CGFloat = 2.5
  var velocityFilterWeight: CGFloat = 0.7
  var minDistance: CGFloat = 5
  var dotSize: CGFloat = 0

  // MARK: - Events

  let onBegin = EventDispatcher()
  let onEnd = EventDispatcher()
  let onInkChange = EventDispatcher()

  // MARK: - Rendering state

  private let imageView = UIImageView()
  private var buffer: CGContext?
  private var bufferSize = CGSize.zero  // in points
  private var bufferScale: CGFloat = 1

  private(set) var isEmptySignature = true
  private var inkBounds: CGRect?

  // MARK: - Stroke state (signature_pad `_reset` / `_lastPoints`)

  private var sampleWindow: [InkSample] = []
  private var lastVelocity: CGFloat = 0
  private var lastWidth: CGFloat = 1.5
  private var isStroking = false

  private struct CurveSegment {
    var startPoint: InkSample
    var control1: CGPoint
    var control2: CGPoint
    var endPoint: InkSample
    var startWidth: CGFloat
    var endWidth: CGFloat
  }

  /**
   * Begins immediately on touch-down (before any movement), so an ancestor
   * UIScrollView's pan recognizer — which needs ~10pt of movement — can never
   * start and steal the stroke mid-draw. `cancelsTouchesInView = false` keeps
   * the raw touches (incl. coalesced samples) flowing to `touchesMoved`.
   */
  private let strokeGuard = UILongPressGestureRecognizer()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    isMultipleTouchEnabled = false
    imageView.contentMode = .topLeft
    imageView.isUserInteractionEnabled = false
    addSubview(imageView)

    strokeGuard.minimumPressDuration = 0
    strokeGuard.allowableMovement = .greatestFiniteMagnitude
    strokeGuard.cancelsTouchesInView = false
    strokeGuard.delaysTouchesBegan = false
    strokeGuard.delaysTouchesEnded = false
    addGestureRecognizer(strokeGuard)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    imageView.frame = bounds
    ensureBuffer()
  }

  // MARK: - Backing buffer

  private func ensureBuffer() {
    let size = bounds.size
    guard size.width >= 1, size.height >= 1 else {
      return
    }
    let scale = max(traitCollection.displayScale, 1)
    if buffer != nil && size == bufferSize && scale == bufferScale {
      return
    }

    let widthPx = max(Int((size.width * scale).rounded()), 1)
    let heightPx = max(Int((size.height * scale).rounded()), 1)
    guard let context = CGContext(
      data: nil,
      width: widthPx,
      height: heightPx,
      bitsPerComponent: 8,
      bytesPerRow: 0,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
    ) else {
      return
    }
    // Flip into UIKit-style coordinates (origin top-left) and draw in points.
    context.translateBy(x: 0, y: CGFloat(heightPx))
    context.scaleBy(x: scale, y: -scale)
    context.setShouldAntialias(true)

    // Preserve existing ink across resizes, anchored top-left, unscaled.
    if let oldImage = buffer?.makeImage() {
      let oldPointSize = CGSize(
        width: CGFloat(oldImage.width) / bufferScale,
        height: CGFloat(oldImage.height) / bufferScale
      )
      context.saveGState()
      context.translateBy(x: 0, y: oldPointSize.height)
      context.scaleBy(x: 1, y: -1)
      context.draw(oldImage, in: CGRect(origin: .zero, size: oldPointSize))
      context.restoreGState()
    }

    buffer = context
    bufferSize = size
    bufferScale = scale
    refreshDisplay()
  }

  private func refreshDisplay() {
    guard let cgImage = buffer?.makeImage() else {
      imageView.image = nil
      return
    }
    imageView.image = UIImage(cgImage: cgImage, scale: bufferScale, orientation: .up)
  }

  // MARK: - Touch handling

  override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard let touch = touches.first else {
      return
    }
    ensureBuffer()
    isStroking = true
    beginStroke()
    onBegin([:])
    process(touch: touch, event: event)
    refreshDisplay()
  }

  override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard isStroking, let touch = touches.first else {
      return
    }
    process(touch: touch, event: event)
    refreshDisplay()
  }

  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard isStroking, let touch = touches.first else {
      return
    }
    process(touch: touch, event: event)
    endStrokeSession()
  }

  override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard isStroking else {
      return
    }
    endStrokeSession()
  }

  private func process(touch: UITouch, event: UIEvent?) {
    let samples = event?.coalescedTouches(for: touch) ?? [touch]
    for sample in samples {
      let location = sample.location(in: self)
      addSample(InkSample(x: location.x, y: location.y, time: sample.timestamp * 1000))
    }
  }

  private func endStrokeSession() {
    isStroking = false
    refreshDisplay()
    onEnd([:])
    onInkChange(["isEmpty": isEmptySignature])
  }

  // MARK: - signature_pad algorithm

  private func beginStroke() {
    sampleWindow.removeAll(keepingCapacity: true)
    lastVelocity = 0
    lastWidth = (minWidth + maxWidth) / 2
  }

  private func addSample(_ point: InkSample) {
    let last = sampleWindow.last
    if let last, distance(from: last, to: point) <= minDistance {
      return
    }
    let curve = pushToWindow(point)
    if last == nil {
      drawDot(x: point.x, y: point.y, radius: dotRadius())
    } else if let curve {
      drawCurve(curve)
    }
  }

  private func pushToWindow(_ point: InkSample) -> CurveSegment? {
    sampleWindow.append(point)
    if sampleWindow.count > 2 {
      // The first curve reuses the first point as its leading neighbor.
      if sampleWindow.count == 3 {
        sampleWindow.insert(sampleWindow[0], at: 0)
      }
      let widths = calculateCurveWidths(start: sampleWindow[1], end: sampleWindow[2])
      let curve = bezierFromWindow(widths: widths)
      sampleWindow.removeFirst()
      return curve
    }
    return nil
  }

  private func calculateCurveWidths(start: InkSample, end: InkSample) -> (start: CGFloat, end: CGFloat) {
    let velocity = velocityFilterWeight * velocity(from: start, to: end)
      + (1 - velocityFilterWeight) * lastVelocity
    let newWidth = max(maxWidth / (velocity + 1), minWidth)
    let widths = (start: lastWidth, end: newWidth)
    lastVelocity = velocity
    lastWidth = newWidth
    return widths
  }

  private func distance(from a: InkSample, to b: InkSample) -> CGFloat {
    let dx = a.x - b.x
    let dy = a.y - b.y
    return sqrt(dx * dx + dy * dy)
  }

  private func velocity(from start: InkSample, to end: InkSample) -> CGFloat {
    return end.time != start.time
      ? distance(from: start, to: end) / CGFloat(end.time - start.time)
      : 0
  }

  private func dotRadius() -> CGFloat {
    return dotSize > 0 ? dotSize : (minWidth + maxWidth) / 2
  }

  private func controlPoints(_ s1: InkSample, _ s2: InkSample, _ s3: InkSample) -> (c1: CGPoint, c2: CGPoint) {
    let dx1 = s1.x - s2.x
    let dy1 = s1.y - s2.y
    let dx2 = s2.x - s3.x
    let dy2 = s2.y - s3.y

    let m1 = CGPoint(x: (s1.x + s2.x) / 2.0, y: (s1.y + s2.y) / 2.0)
    let m2 = CGPoint(x: (s2.x + s3.x) / 2.0, y: (s2.y + s3.y) / 2.0)

    let l1 = sqrt(dx1 * dx1 + dy1 * dy1)
    let l2 = sqrt(dx2 * dx2 + dy2 * dy2)

    let dxm = m1.x - m2.x
    let dym = m1.y - m2.y

    let k: CGFloat = (l1 + l2) == 0 ? 0 : l2 / (l1 + l2)
    let cm = CGPoint(x: m2.x + dxm * k, y: m2.y + dym * k)

    let tx = s2.x - cm.x
    let ty = s2.y - cm.y

    return (
      c1: CGPoint(x: m1.x + tx, y: m1.y + ty),
      c2: CGPoint(x: m2.x + tx, y: m2.y + ty)
    )
  }

  private func bezierFromWindow(widths: (start: CGFloat, end: CGFloat)) -> CurveSegment {
    let c2 = controlPoints(sampleWindow[0], sampleWindow[1], sampleWindow[2]).c2
    let c3 = controlPoints(sampleWindow[1], sampleWindow[2], sampleWindow[3]).c1
    return CurveSegment(
      startPoint: sampleWindow[1],
      control1: c2,
      control2: c3,
      endPoint: sampleWindow[2],
      startWidth: widths.start,
      endWidth: widths.end
    )
  }

  private func bezierPoint(_ t: CGFloat, _ start: CGFloat, _ c1: CGFloat, _ c2: CGFloat, _ end: CGFloat) -> CGFloat {
    return start * (1.0 - t) * (1.0 - t) * (1.0 - t)
      + 3.0 * c1 * (1.0 - t) * (1.0 - t) * t
      + 3.0 * c2 * (1.0 - t) * t * t
      + end * t * t * t
  }

  private func bezierLength(_ curve: CurveSegment) -> CGFloat {
    let steps = 10
    var length: CGFloat = 0
    var px: CGFloat = 0
    var py: CGFloat = 0
    for i in 0...steps {
      let t = CGFloat(i) / CGFloat(steps)
      let cx = bezierPoint(t, curve.startPoint.x, curve.control1.x, curve.control2.x, curve.endPoint.x)
      let cy = bezierPoint(t, curve.startPoint.y, curve.control1.y, curve.control2.y, curve.endPoint.y)
      if i > 0 {
        let xdiff = cx - px
        let ydiff = cy - py
        length += sqrt(xdiff * xdiff + ydiff * ydiff)
      }
      px = cx
      py = cy
    }
    return length
  }

  // MARK: - Incremental drawing

  private func drawCurve(_ curve: CurveSegment) {
    guard let context = buffer else {
      return
    }
    let widthDelta = curve.endWidth - curve.startWidth
    let drawSteps = Int(ceil(bezierLength(curve))) * 2
    guard drawSteps > 0 else {
      return
    }

    let path = CGMutablePath()
    for i in 0..<drawSteps {
      let t = CGFloat(i) / CGFloat(drawSteps)
      let tt = t * t
      let ttt = tt * t
      let u = 1 - t
      let uu = u * u
      let uuu = uu * u

      var x = uuu * curve.startPoint.x
      x += 3 * uu * t * curve.control1.x
      x += 3 * u * tt * curve.control2.x
      x += ttt * curve.endPoint.x

      var y = uuu * curve.startPoint.y
      y += 3 * uu * t * curve.control1.y
      y += 3 * u * tt * curve.control2.y
      y += ttt * curve.endPoint.y

      let width = min(curve.startWidth + ttt * widthDelta, maxWidth)
      path.addEllipse(in: CGRect(x: x - width, y: y - width, width: width * 2, height: width * 2))
      trackBounds(x: x, y: y, radius: width)
    }

    context.setFillColor(penColor.cgColor)
    context.addPath(path)
    context.fillPath()
    markNotEmpty()
  }

  private func drawDot(x: CGFloat, y: CGFloat, radius: CGFloat) {
    guard let context = buffer else {
      return
    }
    let path = CGMutablePath()
    path.addEllipse(in: CGRect(x: x - radius, y: y - radius, width: radius * 2, height: radius * 2))
    context.setFillColor(penColor.cgColor)
    context.addPath(path)
    context.fillPath()
    trackBounds(x: x, y: y, radius: radius)
    markNotEmpty()
  }

  private func trackBounds(x: CGFloat, y: CGFloat, radius: CGFloat) {
    let rect = CGRect(x: x - radius, y: y - radius, width: radius * 2, height: radius * 2)
    inkBounds = inkBounds?.union(rect) ?? rect
  }

  private func markNotEmpty() {
    if isEmptySignature {
      isEmptySignature = false
      onInkChange(["isEmpty": false])
    }
  }

  // MARK: - Commands (invoked from the module on the main thread)

  func clear() {
    sampleWindow.removeAll()
    isStroking = false
    inkBounds = nil
    buffer?.clear(CGRect(origin: .zero, size: bufferSize))
    refreshDisplay()
    if !isEmptySignature {
      isEmptySignature = true
      onInkChange(["isEmpty": true])
    }
  }

  /**
   * Snapshot the ink as a CGImage, optionally trimmed to the ink bounds and
   * composited over a background color. Returns nil when the drawing surface
   * has not been laid out yet.
   */
  func renderExportImage(trim: Bool, backgroundColor: UIColor?) -> CGImage? {
    guard var image = buffer?.makeImage() else {
      return nil
    }

    if trim, let bounds = inkBounds {
      let clamped = bounds.intersection(CGRect(origin: .zero, size: bufferSize))
      if !clamped.isEmpty {
        let pixelRect = CGRect(
          x: clamped.minX * bufferScale,
          y: clamped.minY * bufferScale,
          width: clamped.width * bufferScale,
          height: clamped.height * bufferScale
        ).integral.intersection(CGRect(x: 0, y: 0, width: image.width, height: image.height))
        if !pixelRect.isEmpty, let cropped = image.cropping(to: pixelRect) {
          image = cropped
        }
      }
    }

    guard let background = backgroundColor else {
      return image
    }
    guard let context = CGContext(
      data: nil,
      width: image.width,
      height: image.height,
      bitsPerComponent: 8,
      bytesPerRow: 0,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
    ) else {
      return image
    }
    let rect = CGRect(x: 0, y: 0, width: image.width, height: image.height)
    context.setFillColor(background.cgColor)
    context.fill(rect)
    context.draw(image, in: rect)
    return context.makeImage() ?? image
  }
}

File: android/src/main/java/expo/modules/signatures/SignatureView.kt
// Ink algorithm ported from signature_pad (https://github.com/szimek/signature_pad),
// MIT License, Copyright (c) 2023 Szymon Nowak.
// Keep the math in sync with src/signatureMath.ts — it is validated by the
// golden vectors in test-fixtures/signature-golden.json.

package expo.modules.signatures

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Rect
import android.graphics.RectF
import android.view.MotionEvent
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/** One touch sample in dp coordinates; time in milliseconds. */
data class InkSample(val x: Double, val y: Double, val time: Double)

class SignatureView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  // Ink parameters (set via props). Widths and distances are in dp so the
  // geometry matches iOS points and web CSS pixels.
  var penColor: Int = Color.BLACK
  var minWidth: Double = 0.5
  var maxWidth: Double = 2.5
  var velocityFilterWeight: Double = 0.7
  var minDistance: Double = 5.0
  var dotSize: Double = 0.0

  private val onBegin by EventDispatcher<Unit>()
  private val onEnd by EventDispatcher<Unit>()
  private val onInkChange by EventDispatcher<Map<String, Any>>()

  // Backing buffer: a px-sized bitmap drawn through a density-scaled canvas,
  // so all ink geometry stays in dp.
  private var bitmap: Bitmap? = null
  private var bufferCanvas: Canvas? = null
  private val density: Float = context.resources.displayMetrics.density

  private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val fillPath = Path()

  var isEmptySignature = true
    private set
  private var inkBounds: RectF? = null // dp coordinates

  // Stroke state (signature_pad `_reset` / `_lastPoints`)
  private val sampleWindow = ArrayList<InkSample>(5)
  private var lastVelocity = 0.0
  private var lastWidth = 1.5
  private var isStroking = false

  private class CurveSegment(
    val startPoint: InkSample,
    val c1x: Double,
    val c1y: Double,
    val c2x: Double,
    val c2y: Double,
    val endPoint: InkSample,
    val startWidth: Double,
    val endWidth: Double,
  )

  init {
    setWillNotDraw(false)
  }

  override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
    super.onSizeChanged(w, h, oldw, oldh)
    if (w <= 0 || h <= 0) {
      return
    }
    val old = bitmap
    val next = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
    val canvas = Canvas(next)
    // Preserve existing ink across resizes, anchored top-left, unscaled.
    if (old != null) {
      canvas.drawBitmap(old, 0f, 0f, null)
    }
    canvas.scale(density, density)
    bitmap = next
    bufferCanvas = canvas
    old?.recycle()
    invalidate()
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    bitmap?.let { canvas.drawBitmap(it, 0f, 0f, null) }
  }

  @SuppressLint("ClickableViewAccessibility")
  override fun onTouchEvent(event: MotionEvent): Boolean {
    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        parent?.requestDisallowInterceptTouchEvent(true)
        isStroking = true
        beginStroke()
        onBegin(Unit)
        processMotion(event)
        invalidate()
      }
      MotionEvent.ACTION_MOVE -> {
        if (!isStroking) {
          return false
        }
        processMotion(event)
        invalidate()
      }
      MotionEvent.ACTION_UP -> {
        if (!isStroking) {
          return false
        }
        processMotion(event)
        endStrokeSession()
      }
      MotionEvent.ACTION_CANCEL -> {
        if (!isStroking) {
          return false
        }
        endStrokeSession()
      }
      else -> return false
    }
    return true
  }

  private fun processMotion(event: MotionEvent) {
    for (i in 0 until event.historySize) {
      addSample(
        InkSample(
          (event.getHistoricalX(i) / density).toDouble(),
          (event.getHistoricalY(i) / density).toDouble(),
          event.getHistoricalEventTime(i).toDouble(),
        ),
      )
    }
    addSample(
      InkSample(
        (event.x / density).toDouble(),
        (event.y / density).toDouble(),
        event.eventTime.toDouble(),
      ),
    )
  }

  private fun endStrokeSession() {
    isStroking = false
    invalidate()
    onEnd(Unit)
    onInkChange(mapOf("isEmpty" to isEmptySignature))
  }

  // MARK: signature_pad algorithm

  private fun beginStroke() {
    sampleWindow.clear()
    lastVelocity = 0.0
    lastWidth = (minWidth + maxWidth) / 2
  }

  private fun addSample(point: InkSample) {
    val last = sampleWindow.lastOrNull()
    if (last != null && distance(last, point) <= minDistance) {
      return
    }
    val curve = pushToWindow(point)
    if (last == null) {
      drawDot(point.x, point.y, dotRadius())
    } else if (curve != null) {
      drawCurve(curve)
    }
  }

  private fun pushToWindow(point: InkSample): CurveSegment? {
    sampleWindow.add(point)
    if (sampleWindow.size > 2) {
      // The first curve reuses the first point as its leading neighbor.
      if (sampleWindow.size == 3) {
        sampleWindow.add(0, sampleWindow[0])
      }
      val widths = calculateCurveWidths(sampleWindow[1], sampleWindow[2])
      val cp1 = controlPoints(sampleWindow[0], sampleWindow[1], sampleWindow[2])
      val cp2 = controlPoints(sampleWindow[1], sampleWindow[2], sampleWindow[3])
      val curve = CurveSegment(
        startPoint = sampleWindow[1],
        c1x = cp1[2],
        c1y = cp1[3],
        c2x = cp2[0],
        c2y = cp2[1],
        endPoint = sampleWindow[2],
        startWidth = widths.first,
        endWidth = widths.second,
      )
      sampleWindow.removeAt(0)
      return curve
    }
    return null
  }

  private fun calculateCurveWidths(start: InkSample, end: InkSample): Pair<Double, Double> {
    val velocity = velocityFilterWeight * velocityFrom(start, end) +
      (1 - velocityFilterWeight) * lastVelocity
    val newWidth = max(maxWidth / (velocity + 1), minWidth)
    val widths = Pair(lastWidth, newWidth)
    lastVelocity = velocity
    lastWidth = newWidth
    return widths
  }

  private fun distance(a: InkSample, b: InkSample): Double {
    val dx = a.x - b.x
    val dy = a.y - b.y
    return sqrt(dx * dx + dy * dy)
  }

  private fun velocityFrom(start: InkSample, end: InkSample): Double {
    return if (end.time != start.time) distance(start, end) / (end.time - start.time) else 0.0
  }

  private fun dotRadius(): Double {
    return if (dotSize > 0) dotSize else (minWidth + maxWidth) / 2
  }

  /** Returns [c1x, c1y, c2x, c2y]. */
  private fun controlPoints(s1: InkSample, s2: InkSample, s3: InkSample): DoubleArray {
    val dx1 = s1.x - s2.x
    val dy1 = s1.y - s2.y
    val dx2 = s2.x - s3.x
    val dy2 = s2.y - s3.y

    val m1x = (s1.x + s2.x) / 2.0
    val m1y = (s1.y + s2.y) / 2.0
    val m2x = (s2.x + s3.x) / 2.0
    val m2y = (s2.y + s3.y) / 2.0

    val l1 = sqrt(dx1 * dx1 + dy1 * dy1)
    val l2 = sqrt(dx2 * dx2 + dy2 * dy2)

    val dxm = m1x - m2x
    val dym = m1y - m2y

    val k = if (l1 + l2 == 0.0) 0.0 else l2 / (l1 + l2)
    val cmx = m2x + dxm * k
    val cmy = m2y + dym * k

    val tx = s2.x - cmx
    val ty = s2.y - cmy

    return doubleArrayOf(m1x + tx, m1y + ty, m2x + tx, m2y + ty)
  }

  private fun bezierPoint(t: Double, start: Double, c1: Double, c2: Double, end: Double): Double {
    return start * (1.0 - t) * (1.0 - t) * (1.0 - t) +
      3.0 * c1 * (1.0 - t) * (1.0 - t) * t +
      3.0 * c2 * (1.0 - t) * t * t +
      end * t * t * t
  }

  private fun bezierLength(curve: CurveSegment): Double {
    val steps = 10
    var length = 0.0
    var px = 0.0
    var py = 0.0
    for (i in 0..steps) {
      val t = i.toDouble() / steps
      val cx = bezierPoint(t, curve.startPoint.x, curve.c1x, curve.c2x, curve.endPoint.x)
      val cy = bezierPoint(t, curve.startPoint.y, curve.c1y, curve.c2y, curve.endPoint.y)
      if (i > 0) {
        val xdiff = cx - px
        val ydiff = cy - py
        length += sqrt(xdiff * xdiff + ydiff * ydiff)
      }
      px = cx
      py = cy
    }
    return length
  }

  // MARK: Incremental drawing

  private fun drawCurve(curve: CurveSegment) {
    val canvas = bufferCanvas ?: return
    val widthDelta = curve.endWidth - curve.startWidth
    val drawSteps = ceil(bezierLength(curve)).toInt() * 2
    if (drawSteps <= 0) {
      return
    }

    fillPath.rewind()
    for (i in 0 until drawSteps) {
      val t = i.toDouble() / drawSteps
      val tt = t * t
      val ttt = tt * t
      val u = 1 - t
      val uu = u * u
      val uuu = uu * u

      var x = uuu * curve.startPoint.x
      x += 3 * uu * t * curve.c1x
      x += 3 * u * tt * curve.c2x
      x += ttt * curve.endPoint.x

      var y = uuu * curve.startPoint.y
      y += 3 * uu * t * curve.c1y
      y += 3 * u * tt * curve.c2y
      y += ttt * curve.endPoint.y

      val width = min(curve.startWidth + ttt * widthDelta, maxWidth)
      fillPath.addCircle(x.toFloat(), y.toFloat(), width.toFloat(), Path.Direction.CW)
      trackBounds(x, y, width)
    }

    paint.color = penColor
    canvas.drawPath(fillPath, paint)
    markNotEmpty()
  }

  private fun drawDot(x: Double, y: Double, radius: Double) {
    val canvas = bufferCanvas ?: return
    fillPath.rewind()
    fillPath.addCircle(x.toFloat(), y.toFloat(), radius.toFloat(), Path.Direction.CW)
    paint.color = penColor
    canvas.drawPath(fillPath, paint)
    trackBounds(x, y, radius)
    markNotEmpty()
  }

  private fun trackBounds(x: Double, y: Double, radius: Double) {
    val left = (x - radius).toFloat()
    val top = (y - radius).toFloat()
    val right = (x + radius).toFloat()
    val bottom = (y + radius).toFloat()
    val bounds = inkBounds
    if (bounds == null) {
      inkBounds = RectF(left, top, right, bottom)
    } else {
      bounds.union(left, top, right, bottom)
    }
  }

  private fun markNotEmpty() {
    if (isEmptySignature) {
      isEmptySignature = false
      onInkChange(mapOf("isEmpty" to false))
    }
  }

  // MARK: Commands (invoked from the module on the main thread)

  fun clear() {
    sampleWindow.clear()
    isStroking = false
    inkBounds = null
    bitmap?.eraseColor(Color.TRANSPARENT)
    invalidate()
    if (!isEmptySignature) {
      isEmptySignature = true
      onInkChange(mapOf("isEmpty" to true))
    }
  }

  /**
   * Snapshot the ink into a new bitmap, optionally trimmed to the ink bounds
   * and composited over a background color. Returns null when the drawing
   * surface has not been laid out yet.
   */
  fun renderExportBitmap(trim: Boolean, backgroundColor: Int?): Bitmap? {
    val source = bitmap ?: return null
    var srcRect = Rect(0, 0, source.width, source.height)
    if (trim) {
      inkBounds?.let { bounds ->
        val left = max(floor(bounds.left * density).toInt(), 0)
        val top = max(floor(bounds.top * density).toInt(), 0)
        val right = min(ceil(bounds.right * density).toInt(), source.width)
        val bottom = min(ceil(bounds.bottom * density).toInt(), source.height)
        if (right - left >= 1 && bottom - top >= 1) {
          srcRect = Rect(left, top, right, bottom)
        }
      }
    }
    val output = Bitmap.createBitmap(srcRect.width(), srcRect.height(), Bitmap.Config.ARGB_8888)
    val canvas = Canvas(output)
    if (backgroundColor != null) {
      canvas.drawColor(backgroundColor)
    }
    canvas.drawBitmap(source, srcRect, Rect(0, 0, srcRect.width(), srcRect.height()), null)
    return output
  }
}
